File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Map<Integer, Integer> count = new HashMap<>();
        for (var x : nums2) {
            count.put(x, count.getOrDefault(x, 0)+1);
        }

        ArrayList<Integer> ret = new ArrayList<>();
        for (var x: nums1) {
            if (!count.containsKey(x))
                continue;
            ret.add(x);
            count.put(x, count.get(x)-1);
            if (count.get(x) == 0) {
                count.remove(x);
            }
        }

        return ret.stream().mapToInt(x -> x).toArray();
    }
}

File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        Map2 = Counter(nums2)
        ret = []
        for x in nums1:
            if x not in Map2: 
                continue
            ret.append(x)
            Map2[x] -= 1
            if Map2[x]==0:
                del Map2[x]
        return ret 